marshal/unmarshal.go

```diff
@@ -0,0 +1,142 @@
+package marshal
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+// Unmarshals a JSON element and fills in the ForceSendFields field if
+// the struct contains it. Only anotates basic types in the ForceSendFields.
+func Unmarshal(data []byte, v any) error {
+	if len(data) == 0 {
+		return nil
+	}
+	// create a new element if the pointer is nul
+	if v == nil {
+		objectType := reflect.ValueOf(v).Type()
+		v = reflect.New(objectType).Interface()
+	}
+	err := unmarshalInternal(data, v)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func setForceSendFields(v any, presentFields []string) error {
+
+	value := reflect.ValueOf(v)
+	if value.Kind() != reflect.Ptr {
+		return nil
+	}
+
+	// reflex.GetFieldByName may return the field from an anonymous field
+	// if the extending struct doesn't have the field.
+	field := getFieldByName(v, forceSendFieldName)
+	if !field.IsValid() {
+		return nil
+	}
+
+	if !field.CanSet() || field.Kind() != reflect.Slice {
+		return errors.New("cannot set field")
+	}
+
+	if len(presentFields) == 0 {
+		return nil
+	}
+	presentFieldsValue := reflect.ValueOf(presentFields)
+	field.Set(presentFieldsValue)
+
+	return nil
+}
+
+// Extract the json name from the json tag
+func getJsonName(val string) (string, error) {
+	if val == "-" {
+		return "-", nil
+	}
+
+	i := strings.Index(val, ",")
+
+	if i == -1 {
+		return val, nil
+	}
+	if val[:i] == "" {
+		return "", fmt.Errorf("malformed json tag: %s", val)
+	}
+
+	return val[:i], nil
+}
+
+func unmarshalInternal(data []byte, v any) error {
+	var jsonFields map[string]json.RawMessage
+	err := json.Unmarshal([]byte(data), &jsonFields)
+	if err != nil {
+		return err
+	}
+
+	value := reflect.ValueOf(v)
+	derefer := reflect.Indirect(value)
+
+	objectType := derefer.Type()
+
+	foundFields := []string{}
+	fields := *getTypeFields(objectType)
+
+	for i := 0; i < derefer.NumField(); i++ {
+		field := derefer.Field(i)
+		fieldType := field.Type()
+		jsonTag := fields[i].JsonTag
```

Comment:
We can cache the `jsonTag` object such that parsing/`getJsonName` isn't needed on every call.

Reply:
Added separately. Adding it inside the same cache means that each call to getTypeFields needs to handle errors, which made the code less readable for no real reason.

Reply:
Why would it need to handle errors?

Reply:
Parsing the JSON tags can throw errors, unlike the reflection actions. The auxiliary functions which just want the list of fields and their names don't return errors currently. If we add parsing to the same function, then the function can return an error and those auxiliary functions also need to handle it. 
We can assume that the cache is populated in advance (which should happen) and then ignore errors inside the auxiliary functions, but seems something easy to forget in the future.